staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go

```diff
@@ -101,7 +102,7 @@ func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
 	resp, err := rt.RoundTrip(req)
 
 	if err != nil {
-		return nil, fmt.Errorf("error trying to reach service: %w", err)
+		return nil, errors.NewServiceUnavailable(fmt.Sprintf("error trying to reach service: %v", err))
```

Comment:
Would it be possible to extend `NewServiceUnavailable` so that it could be used with `errors.Is` and `errors.As` method?

I have opened[ a PR](https://github.com/kubernetes/kubernetes/pull/92291) that adds a retry mechanism to the `aggregator`. The idea was to intercept errors returned from the round trippers and interpret them. Here is an example - https://github.com/kubernetes/kubernetes/pull/92291/files?file-filters%5B%5D=.go#diff-ffaa928dbabab1da2e82594be1620b945b37f7fc7256ec1c067abb9af412c4a3R196

Does it make sense ?

Reply:
Got it.  We can extend the [StatusError](https://github.com/kubernetes/kubernetes/blob/838bb6a56760a2abf03bbd42ab3e3a53ed58df13/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L35),
   * adding a new field to record the original error
   * adding helper functions to create StatusError with original error.
   * adding a Unwrap() method

StatusError is used in many places, we should change its interface with caution. We can begin with making a local extension in k8s.io/apimachinery/pkg/util/proxy, and merge it back when it's mature.

Is this PR breaking existing code though? The PR you linked is still open. It would be hard to justify the complexity of the extension if we don't have a real use case.

Reply:
I think that we should follow the standard golang way and add an Unwrap() method.

> Is this PR breaking existing code though? The PR you linked is still open. It would be hard to justify the complexity of the extension if we don't have a real use case.

You are right, this PR won't break existing code it will be just a little harder in the future. I'm okay with merging.